build_rules.gradle

```diff
@@ -1253,6 +1254,125 @@ ext.applyGrpcNature = {
+    relocate "com.google.cloud", "org.apache.beam.vendor.proto_google_common_protos.v1.com.google.cloud"
+    relocate "com.google.logging", "org.apache.beam.vendor.proto_google_common_protos.v1.com.google.logging"
+    relocate "com.google.longrunning", "org.apache.beam.vendor.proto_google_common_protos.v1.com.google.longrunning"
+    relocate "com.google.rpc", "org.apache.beam.vendor.proto_google_common_protos.v1.com.google.rpc"
+    relocate "com.google.type", "org.apache.beam.vendor.proto_google_common_protos.v1.com.google.type"
+    relocate "io.opencensus", "org.apache.beam.vendor.opencensus.v0_11.io.opencensus"
+
+    // Adapted from https://github.com/grpc/grpc-java/blob/e283f70ad91f99c7fee8b31b605ef12a4f9b1690/netty/shaded/build.gradle#L41
+    relocate "io.netty", "org.apache.beam.vendor.netty.v4.io.netty"
+    // We have to be careful with these replacements as they must not match any
+    // string in NativeLibraryLoader, else they cause corruption. Note that
+    // this includes concatenation of string literals and constants.
+    relocate 'META-INF/native/libnetty', 'META-INF/native/liborg_apache_beam_vendor_netty_v4_netty'
+    relocate 'META-INF/native/netty', 'META-INF/native/org_apache_beam_vendor_netty_v4_netty'
+
+    // Don't include errorprone, JDK8 annotations, objenesis, junit, and mockito in the bundled jar
+    exclude "com/google/errorprone/**"
+    exclude "com/google/instrumentation/**"
+    exclude "javax/annotation/**"
+    exclude "junit/**"
+    exclude "org/hamcrest/**"
+    exclude "org/junit/**"
+    exclude "org/mockito/**"
+    exclude "org/objenesis/**"
+  })
+
+  // Don't force modules here because we don't want to take the shared declarations in build_rules.gradle
+  // because we would like to have the freedom to choose which versions of dependencies we
+  // are using for the portability APIs separate from what is being used inside other modules such as GCP.
+  configurations.all { config ->
+    config.resolutionStrategy {
+      forcedModules = []
+    }
+  }
+
+  apply plugin: "com.google.protobuf"
+  protobuf {
+    protoc {
+      // The artifact spec for the Protobuf Compiler
+      artifact = "com.google.protobuf:protoc:3.5.1"
+    }
+
+    // Configure the codegen plugins
+    plugins {
+      // An artifact spec for a protoc plugin, with "grpc" as
+      // the identifier, which can be referred to in the "plugins"
+      // container of the "generateProtoTasks" closure.
+      grpc {
+        artifact = "io.grpc:protoc-gen-grpc-java:1.12.0"
+      }
+    }
+
+    generateProtoTasks {
+      ofSourceSet("main")*.plugins {
+        // Apply the "grpc" plugin whose spec is defined above, without
+        // options.  Note the braces cannot be omitted, otherwise the
+        // plugin will not be added. This is because of the implicit way
+        // NamedDomainObjectContainer binds the methods.
+        grpc { }
+      }
+    }
+  }
+
+  dependencies {
+    compile 'com.google.guava:guava:20.0'
```

Comment:
Is it an explicit choice to maintain these dependency versions separate from the `ext.library` map? Any reason why? It seems like we introduce an opportunity for the versions to drift.

Reply:
With vendoring, version drift only leads to larger jars since we will possibly be packaging multiple versions of the same dependency, other then that it doesn't impact the user. I believe for non-vendored packages, using `ext.library` is the right way to go since we want to have a consistent set of dependencies across the project so users don't have to deal with diamond dependency issues.

In this case I specifically choose different versions because I also wanted to perform a partial upgrade moving us to the latest `protobuf` and `grpc` libraries. It was difficult enough to do this once on a single set of dependencies so I didn't want to repeat this again.

Reply:
Hmm, I still worry about the two version lists drifting. It seems like in the common scenario we want the versions to match, right? Would it be better to use the `ext.library` versions in most cases and only mention explicit overrides where necessary (with a comment about why each is necessary?)

At the very least, I think there should a comment in both source locations giving guidance to maintainers, i.e.

```
applyProtobufNature() {
  // As much as possible, versions here should match those in ext.library. Please add a comment
  // explaining any necessary inconsistencies.
  ..
}

// When updating dependency versions here, please also update vendored versions in applyProtobufNature()
ext.library = {
  // ..
}
```